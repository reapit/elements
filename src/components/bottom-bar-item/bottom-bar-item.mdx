import { Meta, Canvas, Controls } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as BottomBarItemStories from './bottom-bar-item.stories'

<Meta of={BottomBarItemStories} />

# Bottom Bar Item

A versatile component designed to render navigation icon items within a `BottomBar`

`BottomBarItem` is backed by `HTMLAnchorElement` and `HTMLButtonElement` to allow for flexibility in rendering the component

<Controls />

## Default State

The default appearance of the `BottomBarItem`

<Canvas of={BottomBarItemStories.Default} />

## Active State

The visual state of the `BottomBarItem` when it's currently selected or active

<Canvas of={BottomBarItemStories.ActiveState} />

## Badge State

The visual state of the `BottomBarItem` when it's currently selected or active with a badge

<Canvas of={BottomBarItemStories.BadgeState} />

## Style Anchor Usage

Demonstrates how to use the `BottomBarItem` using a standard `HTMLAnchorElement`

<Canvas of={BottomBarItemStories.StyleAnchorUsage} />

<RenderHtmlMarkup of={BottomBarItemStories.StyleAnchorUsage} />

## Style Button Usage

Demonstrates how to use the `BottomBarItem` using a standard `HTMLButtonElement`

<Canvas of={BottomBarItemStories.StyleButtonUsage} />

<RenderHtmlMarkup of={BottomBarItemStories.StyleButtonUsage} />

## React Anchor Usage

Demonstrates how to render the `BottomBarItem` using a anchor version

<Canvas of={BottomBarItemStories.ReactAnchorUsage} />

<RenderHtmlMarkup of={BottomBarItemStories.ReactAnchorUsage} />

## React Button Usage

Demonstrates how to render the `BottomBarItem` using a button version

<Canvas of={BottomBarItemStories.ReactButtonUsage} />

<RenderHtmlMarkup of={BottomBarItemStories.ReactButtonUsage} />
