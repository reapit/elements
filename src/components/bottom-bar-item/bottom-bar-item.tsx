import type { AnchorHTMLAttributes, ButtonHTMLAttributes, FC, MouseEventHandler, ReactNode } from 'react'
import {
  ElAnchorBottomBarItemContainer,
  ElBottomBarItemIcon,
  ElBottomBarItemLabel,
  ElButtonBottomBarItemContainer,
} from './styles'

interface CommonBottomBarItemProps {
  /**
   * The JSX element to render as the icon
   *
   * NOTE: the CSS color property is set to 'inherit' to ensure the icon is aligned with the design
   */
  icon: ReactNode

  /**
   * Render a label for the bottom bar item
   *
   * NOTE: the CSS color property is set to 'inherit' to ensure the label is aligned with the design
   */
  children: ReactNode

  /**
   * Whether the nav item is active
   *
   * @default false
   **/
  isActive?: boolean
}

interface BottomBarItemAsButtonElementProps
  extends CommonBottomBarItemProps,
    Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'children'> {
  href?: never
  onClick: MouseEventHandler<HTMLButtonElement>
}

interface BottomBarItemAsAnchorElementProps
  extends CommonBottomBarItemProps,
    Omit<AnchorHTMLAttributes<HTMLAnchorElement>, 'children'> {
  href?: string
  onClick?: MouseEventHandler<HTMLAnchorElement>
}

export type BottomBarItemProps = BottomBarItemAsButtonElementProps | BottomBarItemAsAnchorElementProps

const isButtonElement = (props: BottomBarItemProps): props is BottomBarItemAsButtonElementProps => {
  return !props?.href
}

/**
 * The `BottomBarItem` component is a small part of `BottomBar` component
 *
 * To ensure it's aligned with the design, the children component should have a CSS color property with 'inherit' value
 */
export const BottomBarItem: FC<BottomBarItemProps> = (props) => {
  if (isButtonElement(props)) {
    const { isActive, icon, children, ...rest } = props ?? {}

    return (
      <ElButtonBottomBarItemContainer {...rest} aria-current={isActive ? 'true' : undefined}>
        <ElBottomBarItemIcon role="presentation">{icon}</ElBottomBarItemIcon>
        <ElBottomBarItemLabel>{children}</ElBottomBarItemLabel>
      </ElButtonBottomBarItemContainer>
    )
  } else {
    const { isActive, icon, children, ...rest } = props ?? {}

    return (
      <ElAnchorBottomBarItemContainer {...rest} aria-current={isActive ? 'page' : undefined}>
        <ElBottomBarItemIcon role="presentation">{icon}</ElBottomBarItemIcon>
        <ElBottomBarItemLabel>{children}</ElBottomBarItemLabel>
      </ElAnchorBottomBarItemContainer>
    )
  }
}
