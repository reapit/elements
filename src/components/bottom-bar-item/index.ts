export * from './bottom-bar-item'
export * from './styles'
