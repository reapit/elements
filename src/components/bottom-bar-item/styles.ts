import { styled } from '@linaria/react'
import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import { ElIcon } from '../icon'

export const ElBottomBarItemIcon = styled.div`
  width: var(--icon-default, 24px);
  height: var(--icon-default, 24px);
  color: inherit;
`

export const ElBottomBarItemLabel = styled.span`
  color: inherit;
  text-align: center;
  font-family: var(--font-family, Inter);
  font-size: var(--font-size-2xs, 12px);
  font-style: normal;
  font-weight: var(--font-weight-regular, Regular);
  line-height: var(--line-height-3xs, 12px);
  letter-spacing: var(--letter-spacing-2xs, 0px);
`

const baseStyles = `
  background-color: var(--fill-white, white);
  outline: none;
  width: 44px; 
  border: var(--border-none, 0);
  display: flex;
  cursor: pointer;
  padding: var(--space-half, 2px) var(--space-none, 0px); 
  flex-direction: column;
  align-items: center;
  gap: var(--space-half, 2px);
  position: relative;

  color: var(--icon-secondary, #607890);

  &:active, &[aria-current="true"], &[aria-current="page"] {
    color: var(--icon-action, #4e56ea)
  }
`

export const ElBottomBarItemBadge = styled.span`
  position: absolute;
  top: 1px;
  right: 6px;
  width: var(--size-2, 8px);
  height: var(--size-2, 8px);
  background-color: var(--icon-error, #f01830);
  border-radius: 100%;
`

export const ElAnchorBottomBarItemContainer = styled.a<AnchorHTMLAttributes<HTMLAnchorElement>>`
  ${baseStyles}

  ${ElIcon} {
    color: inherit;
  }
`

export const ElButtonBottomBarItemContainer = styled.button<ButtonHTMLAttributes<HTMLButtonElement>>`
  ${baseStyles}

  ${ElIcon} {
    color: inherit;
  }
`
