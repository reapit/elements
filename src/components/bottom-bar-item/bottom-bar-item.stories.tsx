import { action } from '@storybook/addon-actions'
import type { Meta, StoryObj } from '@storybook/react'

import { Icon } from '../icon'
import { FlexContainer } from '../layout'
import { BottomBarItem } from './bottom-bar-item'

const meta = {
  title: 'Components/Bottom Bar Item',
  component: BottomBarItem,
  parameters: {
    layout: 'padded',
  },
  argTypes: {
    icon: {
      control: false,
      description: 'The icon to be displayed',
    },
    isActive: {
      control: false,
      description: 'Whether the bottom bar item is active or not',
    },
    hasBadge: {
      control: false,
      description: 'Whether the bottom bar item has a badge or not',
    },
    href: {
      control: false,
      description: 'Specifies the URL for anchor-style bottom bar item',
    },
    onClick: {
      control: false,
      description: 'Click handler for bottom bar item',
    },
    children: {
      control: false,
      description: 'Accessible label for button',
    },
    className: {
      control: false,
      description: 'CSS class for additional styling',
    },
  },
} satisfies Meta<typeof BottomBarItem>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
  args: {
    isActive: false,
    children: 'Label',
    icon: <Icon icon="star" />,
  },
}

export const Active: Story = {
  args: {
    isActive: true,
    children: 'Label',
    icon: <Icon icon="star" />,
  },
}

export const WithBadge: Story = {
  args: {
    children: 'Label',
    icon: <Icon icon="star" />,
    hasBadge: true,
  },
}

export const WithHref: Story = {
  args: {
    href: '#',
    icon: <Icon icon="star" />,
    children: 'Label',
  },
  render: (args) => {
    return (
      <FlexContainer>
        <BottomBarItem {...args} isActive={false} className="el-mr5" />
        <BottomBarItem {...args} isActive className="el-mx5" />
        <BottomBarItem {...args} hasBadge className="el-mx5" />
      </FlexContainer>
    )
  },
}

export const WithOnClick: Story = {
  name: 'With OnClick',
  args: {
    onClick: action('handleClick'),
    icon: <Icon icon="star" />,
    children: 'Label',
  },
  render: (args) => {
    return (
      <FlexContainer>
        <BottomBarItem {...args} isActive={false} className="el-mr5" />
        <BottomBarItem {...args} isActive className="el-mx5" />
        <BottomBarItem {...args} hasBadge className="el-mx5" />
      </FlexContainer>
    )
  },
}
