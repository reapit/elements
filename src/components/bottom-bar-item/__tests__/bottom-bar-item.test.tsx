import { render } from '@testing-library/react'
import { BottomBarItem } from '../bottom-bar-item'

describe('BottomBarItem', () => {
  it('should render the component as HTMLAnchorElement with default state', () => {
    expect(
      render(
        <BottomBarItem href="#" icon={<span>icon</span>}>
          Label
        </BottomBarItem>,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should render the component as HTMLAnchorElement with active state', () => {
    expect(
      render(
        <BottomBarItem href="#" isActive icon={<span>icon</span>}>
          Label
        </BottomBarItem>,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should render the component as HTMLButtonElement with default state', () => {
    expect(
      render(
        <BottomBarItem onClick={vi.fn()} isActive icon={<span>icon</span>}>
          Label
        </BottomBarItem>,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should render the component as HTMLButtonElement with active state', () => {
    expect(
      render(
        <BottomBarItem onClick={vi.fn()} isActive icon={<span>icon</span>}>
          Label
        </BottomBarItem>,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should render the component as HTMLButtonElement with badge state', () => {
    expect(
      render(
        <BottomBarItem onClick={vi.fn()} isActive icon={<span>icon</span>} hasBadge>
          Label
        </BottomBarItem>,
      ).asFragment(),
    ).toMatchSnapshot()
  })
})
