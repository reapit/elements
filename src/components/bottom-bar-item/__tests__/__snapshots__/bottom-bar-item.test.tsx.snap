// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomBarItem should render the component as HTMLAnchorElement with active state 1`] = `
<DocumentFragment>
  <mock-styled.a
    aria-current="page"
    href="#"
  >
    <mock-styled.div
      role="presentation"
    >
      <span>
        icon
      </span>
    </mock-styled.div>
    <mock-styled.span>
      Label
    </mock-styled.span>
  </mock-styled.a>
</DocumentFragment>
`;

exports[`BottomBarItem should render the component as HTMLAnchorElement with default state 1`] = `
<DocumentFragment>
  <mock-styled.a
    href="#"
  >
    <mock-styled.div
      role="presentation"
    >
      <span>
        icon
      </span>
    </mock-styled.div>
    <mock-styled.span>
      Label
    </mock-styled.span>
  </mock-styled.a>
</DocumentFragment>
`;

exports[`BottomBarItem should render the component as HTMLButtonElement with active state 1`] = `
<DocumentFragment>
  <mock-styled.button
    aria-current="true"
  >
    <mock-styled.div
      role="presentation"
    >
      <span>
        icon
      </span>
    </mock-styled.div>
    <mock-styled.span>
      Label
    </mock-styled.span>
  </mock-styled.button>
</DocumentFragment>
`;

exports[`BottomBarItem should render the component as HTMLButtonElement with default state 1`] = `
<DocumentFragment>
  <mock-styled.button
    aria-current="true"
  >
    <mock-styled.div
      role="presentation"
    >
      <span>
        icon
      </span>
    </mock-styled.div>
    <mock-styled.span>
      Label
    </mock-styled.span>
  </mock-styled.button>
</DocumentFragment>
`;
